arangod/IResearch/IResearchDataStore.cpp

```diff
@@ -828,36 +963,79 @@ Result IResearchDataStore::commitUnsafeImpl(
-          << "no changes registered for ArangoSearch index '" << id()
-          << "' got last operation tick '" << _lastCommittedTick << "'";
-
-      // no changes, can release the latest tick before commit
-      impl.tick(lastTickBeforeCommit);
-      _lastCommittedTick = lastTickBeforeCommit;
+#endif
+    _commitStageOne = true;
+    auto stageOneGuard = absl::Cleanup{
+        [&, lastCommittedTickOne = _lastCommittedTickOne]() noexcept {
+          _lastCommittedTickOne = lastCommittedTickOne;
+        }};
+    auto const lastTickBeforeCommitOne = _engine->currentTick();
+    auto const commitOne = _dataStore._writer->commit(progress);
+    std::move(stageOneGuard).Cancel();
+    if (!commitOne) {
+      LOG_TOPIC("7e319", TRACE, TOPIC)
+          << "First commit for ArangoSearch index '" << id()
+          << "' is no changes tick " << lastTickBeforeCommitOne;
+      TRI_ASSERT(_lastCommittedTickOne <= lastTickBeforeCommitOne);
+      TRI_ASSERT(_lastCommittedTickTwo <= lastTickBeforeCommitOne);
+      // no changes, can release the latest tick before commit one
+      _lastCommittedTickOne = lastTickBeforeCommitOne;
+      _lastCommittedTickTwo = lastTickBeforeCommitOne;
+      impl.tick(_lastCommittedTickOne);
       return {};
+    } else {
+      code = CommitResult::DONE;
+    }
+#if ARANGODB_ENABLE_MAINTAINER_MODE && ARANGODB_ENABLE_FAILURE_TESTS
+    TRI_IF_FAILURE("ArangoSearch::ThreeTransactionsMisorder::StageOneKill") {
+      std::unique_lock sync(_t3FailureSync);
+      // if all candidates gathered and max tick is committed it's time to crash
+      if (_t3Candidates.size() >= 3 &&
+          std::find_if(_t3Candidates.begin(), _t3Candidates.end(),
+                       [this](uint64_t t) {
+                         return t > _lastCommittedTickOne;
+                       }) == _t3Candidates.end()) {
+        TRI_TerminateDebugging("Killed on commit stage one");
+      }
+    }
+#endif
+    // now commit what has possibly accumulated in the second flush context
+    // but won't move the tick as it possibly contains "3rd" generation changes
+    _commitStageOne = false;
+    auto stageTwoGuard = absl::Cleanup{
+        [&, lastCommittedTickTwo = _lastCommittedTickTwo]() noexcept {
+          _lastCommittedTickTwo = lastCommittedTickTwo;
+        }};
+    auto const lastTickBeforeCommitTwo = _engine->currentTick();
+    auto const commitTwo = _dataStore._writer->commit(progress);
+    std::move(stageTwoGuard).Cancel();
+    if (!commitTwo) {
+      LOG_TOPIC("21bda", TRACE, TOPIC)
+          << "Second commit for ArangoSearch index '" << id()
+          << "' is no changes tick " << lastTickBeforeCommitTwo;
+      TRI_ASSERT(_lastCommittedTickOne <= lastTickBeforeCommitTwo);
+      TRI_ASSERT(_lastCommittedTickTwo <= lastTickBeforeCommitTwo);
+      // no changes, can release the latest tick before commit two
+      _lastCommittedTickOne = lastTickBeforeCommitTwo;
+      _lastCommittedTickTwo = lastTickBeforeCommitTwo;
```

Comment:
Github is fooling me or failpoint for crash after stage two is missing? https://github.com/arangodb/arangodb/blob/a4eb4d658241b53f3ede34f983ba31f74609a00d/arangod/IResearch/IResearchLink.cpp#L1163

Reply:
True